lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -2676,7 +2676,14 @@ bool LLVMToSPIRV::transGlobalVariables() {
   for (auto I = M->global_begin(), E = M->global_end(); I != E; ++I) {
     if ((*I).getName() == "llvm.global.annotations")
       transGlobalAnnotation(&(*I));
-    else if (MDNode *IO = ((*I).getMetadata("io_pipe_id")))
+    else if ((I->getName() == "llvm.global_ctors" ||
```

Comment:
Am I right that the translator handles these variables as follows:
- if SPIR-V 1.1 is enabled, emit `Initializer`/`Finalizer`
- else if function pointers are available, then translate variables as-is using corresponding extensions
- otherwise, ignore them

right?

Reply:
Almost.  After this PR, the state will be like this:

LLVM -> SPIR-V:
 * if function pointers are available, then translate variables as-is using corresponding extensions
 * regardless of function pointers, if SPIR-V 1.1 is enabled, emit `Initializer` (emitting `Finalizer` is not part of this PR)
 * else report error

SPIR-V -> LLVM:
 * if the SPIR-V does not contain a `llvm.global_ctors` variable, construct it in LLVM IR from `Initializer` entry points.

Does that make sense?

Reply:
> else report error

Seems like this happens only if user explicitly stated that function should be emitted as `Initializer` through corresponding metadata from SPIR-V friendly IR, right? I.e. `llvm.global_ctors` is being silently ignored if function pointers are not available.

> Does that make sense?

Yes, looks good to me

Reply:
> I.e. llvm.global_ctors is being silently ignored if function pointers are not available.

`llvm-spirv --spirv-ext=+SPV_INTEL_function_pointers --spirv-max-version=1.0` will also attempt to emit `Initializer`, so it will fail in the same way as without the `--spirv-ext...` option.  That is because the `Initializer` mode is set by the `PreprocessMetadata` pass early in the translator.  I can relax that if you need `llvm.global_ctors` with function pointers before SPIR-V 1.1?

Reply:
> I can relax that if you need llvm.global_ctors with function pointers before SPIR-V 1.1?

It seems like a nice addition, but I'm not really asking to do that, because I'm not sure that we have such use-cases at the moment. Probably @hewj03 can comment based on #632